lib/AtomicCache.php

```diff
@@ -0,0 +1,283 @@
+    {
+        return call(function () use ($key, $create, $ttl): \Generator {
+            $value = yield $this->cache->get($key);
+
+            if ($value === null) {
+                return null;
+            }
+
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                // Attempt to get the value again, since it may have been set while obtaining the lock.
+                $value = yield $this->cache->get($key);
+
+                if ($value === null) {
+                    return null;
+                }
+
+                return yield from $this->create($create, $key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    /**
+     * The lock is obtained for the key before setting the value.
+     *
+     * @param string   $key   Cache key.
+     * @param mixed    $value Value to cache.
+     * @param int|null $ttl   Timeout in seconds. The default `null` $ttl value indicates no timeout.
+     *
+     * @return Promise<void> Resolves either successfully or fails with a CacheException on failure.
+     *
+     * @throws CacheException
+     * @throws SerializationException
+     *
+     * @see SerializedCache::set()
+     */
+    public function set(string $key, $value, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $value, $ttl): \Generator {
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                yield $this->cache->set($key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+        });
+    }
+
+    /**
+     * The lock is obtained for the key and the value is set only if the key does not exist.
+     *
+     * @param string   $key   Cache key.
+     * @param mixed    $value Value to cache.
+     * @param int|null $ttl   Timeout in seconds. The default `null` $ttl value indicates no timeout.
+     *
+     * @return Promise<mixed> Resolves either with the current value or with the newly set value. Fails with a
+     * CacheException or SerializationException on failure.
+     *
+     * @throws CacheException
+     * @throws SerializationException
+     */
+    public function setIfAbsent(string $key, $value, ?int $ttl = null): Promise
+    {
+        return call(function () use ($key, $value, $ttl): \Generator {
+            $lock = yield from $this->lock($key);
+            \assert($lock instanceof Lock);
+
+            try {
+                $currentValue = yield $this->cache->get($key);
+
+                if ($currentValue !== null) {
+                    return $currentValue;
+                }
+
+                yield $this->cache->set($key, $value, $ttl);
+            } finally {
+                $lock->release();
+            }
+
+            return $value;
+        });
+    }
```

Comment:
Should we also remove this method for now? If one really needs it, `computeIfAbsent` can be used.

Reply:
Sure, I have no use for this method right now.